File: content_publisher/capabilities/planning.py
from __future__ import annotations

from .utils import normalize
from ..integrations.keyword_overlap import evaluate_overlap
from ..interfaces import OpportunityRecord
from ..models import Atom, Cluster, Site, SiteSnapshot, new_id, utc_now


class PlanningCapability:
    def detect_demand(self, snapshot: SiteSnapshot, demand_sources: list) -> list[OpportunityRecord]:
        opportunities: dict[tuple[str, str], OpportunityRecord] = {}
        for source in demand_sources:
            for record in source.collect(snapshot):
                key = (normalize(record.topic), normalize(record.cluster_name))
                current = opportunities.get(key)
                if current is None or record.demand_score > current.demand_score:
                    opportunities[key] = record
        return list(opportunities.values())

    def analyze_gaps(self, snapshot: SiteSnapshot, opportunities: list[OpportunityRecord]) -> list[OpportunityRecord]:
        existing_topics = [atom.topic for atom in snapshot.atoms.values()]
        normalized_topics = {normalize(topic) for topic in existing_topics}
        gaps: list[OpportunityRecord] = []
        for opportunity in opportunities:
            if normalize(opportunity.topic) in normalized_topics:
                continue
            overlap = evaluate_overlap(opportunity.topic, existing_topics)
            if overlap["should_skip"]:
                continue
            gaps.append(opportunity)
        return gaps

    def plan_clusters(self, snapshot: SiteSnapshot, gaps: list[OpportunityRecord]) -> list[Cluster]:
        clusters_by_name = {normalize(cluster.name): cluster for cluster in snapshot.clusters.values()}
        touched: list[Cluster] = []
        for gap in gaps:
            key = normalize(gap.cluster_name)
            cluster = clusters_by_name.get(key)
            if cluster is None:
                cluster = Cluster(
                    id=new_id("cluster"),
                    site_id=snapshot.site.id,
                    name=gap.cluster_name,
                    description=f"Strategic coverage area for {gap.cluster_name}",
                    priority=min(100, gap.demand_score),
                )
                snapshot.clusters[cluster.id] = cluster
                snapshot.site.cluster_ids.append(cluster.id)
                clusters_by_name[key] = cluster
            else:
                cluster.priority = max(cluster.priority, gap.demand_score)
                cluster.updated_at = utc_now()
            if cluster not in touched:
                touched.append(cluster)
        return touched

    def create_atoms(self, snapshot: SiteSnapshot, clusters: list[Cluster], gaps: list[OpportunityRecord]) -> list[Atom]:
        cluster_by_name = {normalize(cluster.name): cluster for cluster in clusters}
        created: list[Atom] = []
        for gap in gaps:
            cluster = cluster_by_name[normalize(gap.cluster_name)]
            atom = Atom(
                id=new_id("atom"),
                site_id=snapshot.site.id,
                cluster_id=cluster.id,
                topic=gap.topic,
                search_intent=gap.search_intent,
                context={
                    "brand_tone": snapshot.site.brand_tone,
                    "target_audience": snapshot.site.target_audience,
                    "monetization_strategy": snapshot.site.monetization_strategy,
                    "demand_score": gap.demand_score,
                    "cluster_name": gap.cluster_name,
                },
                priority=gap.demand_score,
                source_refs=[
                    {
                        "stage": "demand_detection",
                        "source": gap.source,
                        "confidence": gap.confidence,
                        "demand_score": gap.demand_score,
                        "overlap": evaluate_overlap(gap.topic, [existing.topic for existing in snapshot.atoms.values()]),
                    }
                ],
            )
            snapshot.atoms[atom.id] = atom
            snapshot.site.atom_ids.append(atom.id)
            cluster.atom_ids.append(atom.id)
            cluster.updated_at = utc_now()
            created.append(atom)
        return created

File: src/lib/engine/planning.ts
import { Atom, Cluster, OpportunityRecord, SiteSnapshot, newId, utcNow } from "@/domain/models";
import { normalize, overlapSignal } from "@/lib/engine/utils";

export function detectDemand(snapshot: SiteSnapshot): OpportunityRecord[] {
  const pool = snapshot.site.opportunityPool;
  const opportunities = new Map<string, OpportunityRecord>();

  for (const raw of pool) {
    const record: OpportunityRecord = {
      topic: String(raw.topic ?? "").trim(),
      clusterName: String(raw.cluster_name ?? raw.clusterName ?? raw.topic ?? "general").trim(),
      searchIntent: String(raw.search_intent ?? raw.searchIntent ?? "informational").trim(),
      demandScore: Number(raw.demand_score ?? raw.demandScore ?? 50),
      source: String(raw.source ?? "seed").trim(),
      confidence: Number(raw.confidence ?? 0.7)
    };
    if (!record.topic) {
      continue;
    }
    const key = `${normalize(record.topic)}::${normalize(record.clusterName)}`;
    const current = opportunities.get(key);
    if (!current || record.demandScore > current.demandScore) {
      opportunities.set(key, record);
    }
  }

  return [...opportunities.values()];
}

export function analyzeGaps(snapshot: SiteSnapshot, opportunities: OpportunityRecord[]): OpportunityRecord[] {
  const existingTopics = Object.values(snapshot.atoms).map((atom) => atom.topic);
  const normalizedTopics = new Set(existingTopics.map(normalize));

  return opportunities.filter((opportunity) => {
    if (normalizedTopics.has(normalize(opportunity.topic))) {
      return false;
    }
    return !overlapSignal(opportunity.topic, existingTopics).shouldSkip;
  });
}

export function planClusters(snapshot: SiteSnapshot, gaps: OpportunityRecord[]): Cluster[] {
  const byName = new Map<string, Cluster>();
  for (const cluster of Object.values(snapshot.clusters)) {
    byName.set(normalize(cluster.name), cluster);
  }

  const touched: Cluster[] = [];
  for (const gap of gaps) {
    const key = normalize(gap.clusterName);
    let cluster = byName.get(key);
    if (!cluster) {
      cluster = {
        id: newId("cluster"),
        siteId: snapshot.site.id,
        name: gap.clusterName,
        description: `Strategic coverage area for ${gap.clusterName}`,
        priority: Math.min(100, gap.demandScore),
        atomIds: [],
        createdAt: utcNow(),
        updatedAt: utcNow()
      };
      snapshot.clusters[cluster.id] = cluster;
      snapshot.site.clusterIds.push(cluster.id);
      byName.set(key, cluster);
    } else {
      cluster.priority = Math.max(cluster.priority, gap.demandScore);
      cluster.updatedAt = utcNow();
    }
    if (!touched.find((candidate) => candidate.id === cluster.id)) {
      touched.push(cluster);
    }
  }

  return touched;
}

export function createAtoms(snapshot: SiteSnapshot, clusters: Cluster[], gaps: OpportunityRecord[]): Atom[] {
  const clusterByName = new Map(clusters.map((cluster) => [normalize(cluster.name), cluster]));
  const created: Atom[] = [];

  for (const gap of gaps) {
    const cluster = clusterByName.get(normalize(gap.clusterName));
    if (!cluster) {
      continue;
    }
    const atom: Atom = {
      id: newId("atom"),
      siteId: snapshot.site.id,
      clusterId: cluster.id,
      topic: gap.topic,
      searchIntent: gap.searchIntent,
      context: {
        brandTone: snapshot.site.brandTone,
        targetAudience: snapshot.site.targetAudience,
        monetizationStrategy: snapshot.site.monetizationStrategy,
        demandScore: gap.demandScore,
        clusterName: gap.clusterName
      },
      priority: gap.demandScore,
      state: "planned",
      sourceRefs: [
        {
          stage: "demand_detection",
          source: gap.source,
          confidence: gap.confidence,
          demandScore: gap.demandScore
        }
      ],
      outputIds: [],
      signalIds: [],
      insightIds: [],
      createdAt: utcNow(),
      updatedAt: utcNow()
    };
    snapshot.atoms[atom.id] = atom;
    snapshot.site.atomIds.push(atom.id);
    cluster.atomIds.push(atom.id);
    cluster.updatedAt = utcNow();
    created.push(atom);
  }

  return created;
}
